File: 0283-move-zeroes/0283-move-zeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        l = 0
        for r in range(n):
            if nums[l] == 0 and nums[r] != 0:
                # swap
                nums[l],nums[r] = nums[r],nums[l]
                l += 1
            if nums[l] != 0 and nums[r] != 0:
                l += 1


        

# 0     1   0   3   12
# l,r
# l     r
# 1     0   0   3   12 ->swap
#       l   r
#       l       r
# 1     3   0   0   12 ->swap
#           l       r
# 1     3   12  0   0-> swap
            

        

File: 0283-move-zeroes/0283-move-zeroes.cpp
class Solution {
public:
    void moveZeroes(vector<int>& arr) {
        int n = arr.size();
        // using two pointer
        int left = 0;
        for(int right=0; right<n; right++){
            if(arr[left] == 0 && arr[right] != 0){
                swap(arr[left], arr[right]);
                left += 1;
            }
            if(arr[left] != 0){
                left += 1;
            }
        }
    }
};